### ISSUE TITLE
IQueryableExtensions.ToParameterizedSql Assumes DateTime Parameters are Always DateTime and not DateTime2

### ISSUE THREAD

**Issue by paulsavides** [borisdj/EFCore.BulkExtensions#362]
Hello,

Here is a self-contained reproduction of the issue >> https://github.com/paulsavides/DateTimeTesting (Just using LocalDB)

The alternative title to this issue could be `SqlParameter` defaults are not always correct. But, the issue title is a bit more specific to what I'm seeing.

**The Problem**
Somewhere in my application I am executing code something like:
```csharp
var minRecordCreateTime = await context.Entities.Where(e => e.SomeCondition).Select(e => e.CreateTime).FirstOrDefaultAsync();
await context.Entities.Where(e.CreateDate < minRecordCreateTime).BatchDeleteAsync();
```

The CreateDate column is of type `datetime2(7)`. However, in `IQueryableExtensions.ToParameterizedSql` the code will allow `SqlParameter` to select whatever type it finds is most appropriate. In the case of `System.DateTime`, `SqlParameter` will map that to `SqlDbType.DateTime` instead of `SqlDbType.DateTime2`.

Because sqlserver datetime is lower precision than sql server datetime2(7), when the query is actually executed, I end up with rounding errors causing the datetime to always be lower or greater than I expected. Sometimes causing the `minRecordCreateTime` record to be deleted.

Of course, this could generally impact any scenario where the `SqlParameter` default is lower precision than the actual data type in the database.

**Potential Code Fixes**
Debugging through `IQueryableExtensions` I did notice that there is a private property on the implementation of IRelationalParameter that stored what EFCore knows about the actual database type.

This could potentially be fixed by grabbing that private property and always creating the `SqlParameter` with the most specific DbType possible.

Thank you for looking at this issue and for the library. Let me know if you require any clarification.

**Comment by borisdj** [comment]
I'm not sure how to get and set most specific DbType from Relational ParameterValues.
If you manage to implement proposed solution you can send PullRequest.

**Comment by borisdj** [comment]
Fixed with v3.4.3
